mixins/visibility-mixin.js

```diff
@@ -0,0 +1,207 @@
+import { dedupeMixin } from '@open-wc/dedupe-mixin';
+
+const reduceMotion = matchMedia('(prefers-reduced-motion: reduce)').matches;
+const transitionDuration = 300;
+const moveYValue = 10;
+
+export const VisibilityMixin = dedupeMixin(superclass => class extends superclass {
+	static get properties() {
+		return {
+			animate: { type: String }
+		};
+	}
+
+	constructor() {
+		super();
+		this.dummy = document.createElement('div');
+		this.dummy.style.overflow = 'hidden';
+		this.dummy.style.display = 'grid';
+	}
+
+	firstUpdated() {
+		this.opacityOriginal = window.getComputedStyle(this).opacity;
+		this.transformOriginal = window.getComputedStyle(this).transform;
+		this.displayOriginal = window.getComputedStyle(this).display;
+		this.transitionOriginal = window.getComputedStyle(this).transition;
+	}
+
+	updated(changedProperties) {
+		super.updated(changedProperties);
+		if (changedProperties.has('animate')) {
+			if (this.animate === 'show') {
+				this._animateShow();
+			} else if (this.animate === 'hide') {
+				this._animateHide();
+			} else if (this.animate === 'remove') {
+				this._animateRemove();
+			}
+		}
+	}
+
+	_animateHide() {
+		const thisOnTransitionStart = () => {
+			this.dispatchEvent(new CustomEvent(
+				'd2l-visibility-mixin-hide-start',
+				{ bubbles: true, composed: false }
+			));
+		};
+		const dummyOnTransitionEnd = () => {
+			this.dummy.replaceWith(this);
+			this.style.display = 'none';
+			this.dispatchEvent(new CustomEvent(
+				'd2l-visibility-mixin-hide-end',
+				{ bubbles: true, composed: false }
+			));
+		};
+		this._animateHideRemove(thisOnTransitionStart, dummyOnTransitionEnd);
+	}
+
+	_animateHideRemove(thisOnTransitionStart, dummyOnTransitionEnd) {
+		const animateHideRemoveStyle = {
+			initial: {
+				transition: `all ${transitionDuration}ms ease`,
+				opacity: `${this.opacityOriginal}`,
+				transform: `${this.transformOriginal}`
+			},
+			initialDummy: {
+				transition: `height ${transitionDuration}ms ease ${transitionDuration / 3}ms`,
+				height: `calc(${this.scrollHeight}px + ${window.getComputedStyle(this).marginTop} + ${window.getComputedStyle(this).marginBottom})`
+			},
+			final: {
+				opacity: '0',
+				transform: `translateY(-${moveYValue}px)`
+			},
+			finalDummy: {
+				height: '0'
+			},
+			thisOnTransitionStart: thisOnTransitionStart,
+			dummyOnTransitionEnd: dummyOnTransitionEnd
+		};
+		this._animateVisibility(animateHideRemoveStyle);
+	}
+
+	_animateRemove() {
+		const thisOnTransitionStart = () => {
+			this.dispatchEvent(new CustomEvent(
+				'd2l-visibility-mixin-remove-start',
+				{ bubbles: true, composed: false }
+			));
+		};
+		const dummyOnTransitionEnd = () => {
+			this.dummy.replaceWith(this);
+			this.dispatchEvent(new CustomEvent(
+				'd2l-visibility-mixin-remove-end',
+				{ bubbles: true, composed: false }
+			));
+			this.remove();
```

Comment:
I feel like this should come before the event dispatch. However, when I do so, the test checking for 'd2l-visibility-mixin-remove-end' event firing fails. It would say `event.target` is null (and I think `event` itself was null as well). But adding an event listener to it and using `console.log` shows that both `event` and `event.target` are not null.

Reply:
Interesting, although not entirely surprising. It's unfortunate that we need `this.dummy.replaceWith(this)` (which will cause a full re-render) just to dispatch the event and then right after we remove it (causing another re-render). Wondering how important those events are beyond tests? Or can we just not do the `replaceWith` call and have `this` just be a child of `dummy` when the event fires?

Reply:
We could have `this` be a child of `dummy` when the event fires, and then do a `this.dummy.remove()` to get rid of both `this` and `dummy`. Seems to work without any issues, aside from it being strange to have the event dispatch before the actual remove happen.